File: office_convert/probe_lite.py
"""Metadata-only probe that avoids loading the full document into Aspose.

Aspose's per-product probe (`Document(path).get_PageCount()`) loads the
entire input to compute page count. For inputs over ~50 MB, that pushes
the worker past the 2 GB RLIMIT_AS ceiling and the kernel SIGKILLs the
probe (exit 137). The orchestrator never gets a page count, never plans
chunks, never gets a chance to subdivide — the request fails outright.

This module sidesteps Aspose for the probe step:

  - **OOXML (docx/pptx)**: open the .zip and read a few KB from
    `docProps/app.xml` (DOCX page count, PPTX slide count). Memory
    cost: tens of KB regardless of input size. XLSX is intentionally
    excluded: the C++ XLSX probe needs the rendered PDF page count
    (not the worksheet count) for chunk planning, so it always falls
    through to the Aspose worker.

  - **PDF**: shell out to qpdf `--show-npages` (qpdf is already in the
    runtime image for the streaming concat). Uses qpdf's own xref-table
    parser, which reads only the trailer + xref — memory cost bounded
    independent of PDF size.

The price: app.xml is written at *save time* by the authoring application.
If a document was edited externally between save and conversion (rare for
the usual upload flow), the count could drift. For chunk *planning* this
is acceptable — chunks are then merged via qpdf based on actual rendered
output, not a planning count. The render path validates against the live
document.

Returns `ProbeResult | None`. `None` means "no reliable metadata count
extractable; fall through to the C++ Aspose worker." Exceptions from
malformed inputs are swallowed (return None) — the caller's fallback path
is the canonical failure surface.
"""

from __future__ import annotations

import asyncio
import logging
import re
import xml.etree.ElementTree as ET
import zipfile
from pathlib import Path

from office_convert.types import FormatName, ProbeResult

log = logging.getLogger(__name__)

_OOXML_APP_NS = "http://schemas.openxmlformats.org/officeDocument/2006/extended-properties"


async def probe_lite(input_path: Path, format: FormatName) -> ProbeResult | None:
    """Best-effort metadata-only probe. Returns None on any failure."""
    size = input_path.stat().st_size
    try:
        if format == "docx":
            n = await asyncio.to_thread(_ooxml_count, input_path, "Pages")
            # If app.xml doesn't have a page count, use a size-based estimate
            # rather than falling through to the expensive Aspose probe (which
            # loads the entire document for a full layout pass — 10+ minutes
            # for large files). The estimate is conservative (assumes ~20 KB
            # per page for DOCX) so the planner over-chunks slightly; the
            # OOM subdivision path handles any miscalculation.
            if n is None or n <= 0:
                n = _estimate_pages_from_size(size, format)
        elif format == "pptx":
            # Try app.xml first, then count actual slide XML files in the ZIP
            # (always accurate — each slide is a separate XML entry).
            n = await asyncio.to_thread(_ooxml_count, input_path, "Slides")
            if n is None or n <= 0:
                n = await asyncio.to_thread(_pptx_slide_count_from_zip, input_path)
            if n is None or n <= 0:
                n = _estimate_pages_from_size(size, format)
        elif format == "xlsx":
            # XLSX worksheet count was a valid over-estimate under the single-
            # chunk carve-out (one worksheet → at least one PDF page). Once
            # xlsx.cpp adopted PdfSaveOptions::SetPageIndex/SetPageCount, the
            # planner subdivides on *rendered* page count — a 1M-row sheet
            # paginates into thousands of PDF pages, so the worksheet count
            # would silently under-count and the slice would drop most rows.
            # Fall through to the C++ probe (WorkbookRender::GetPageCount).
            return None
        elif format == "pdf":
            n = await _qpdf_page_count(input_path)
        else:
            return None
    except Exception as e:
        log.debug("probe_lite(%s) failed, falling back: %s", format, e)
        return None

    if n is None or n <= 0:
        return None
    return ProbeResult(page_count=n, format=format, natural_seams=(), size_bytes=size)


def _ooxml_count(path: Path, element: str) -> int | None:
    """Read `<Pages>` or `<Slides>` from `docProps/app.xml` inside an OOXML zip.

    Returns None if the file isn't a valid OOXML zip, app.xml is missing,
    or the element isn't present / isn't an integer.
    """
    with zipfile.ZipFile(path) as z:
        try:
            with z.open("docProps/app.xml") as f:
                tree = ET.parse(f)
        except KeyError:
            return None
        node = tree.getroot().find(f"{{{_OOXML_APP_NS}}}{element}")
        if node is None or node.text is None:
            return None
        try:
            return int(node.text.strip())
        except ValueError:
            return None


def _pptx_slide_count_from_zip(path: Path) -> int | None:
    """Count actual slide XML files in the PPTX ZIP structure.

    Each slide in a PPTX is stored as `ppt/slides/slide<N>.xml`. This gives
    the exact slide count without loading the presentation into Aspose —
    just a ZIP directory listing (microseconds, zero memory).

    Returns None if the file isn't a valid ZIP or has no slide entries.
    """
    try:
        with zipfile.ZipFile(path) as z:
            count = sum(
                1 for name in z.namelist()
                if name.startswith("ppt/slides/slide") and name.endswith(".xml")
            )
            return count if count > 0 else None
    except (zipfile.BadZipFile, OSError):
        return None


_QPDF_NPAGES_RE = re.compile(rb"^\s*(\d+)\s*$")


# Conservative bytes-per-page estimates for size-based fallback probing.
# These are intentionally LOW (meaning we estimate MORE pages than reality)
# so the chunk planner produces slightly more chunks than needed. Over-chunking
# is safe (just slightly more subprocess spawns); under-chunking risks OOM.
# Derived from empirical observation across typical Office documents.
_BYTES_PER_PAGE_ESTIMATE: dict[FormatName, int] = {
    "docx": 20_000,   # ~20 KB/page (text-heavy docs are ~5-10 KB, image-heavy ~50-100 KB)
    "pptx": 100_000,  # ~100 KB/slide (slides have images/shapes)
    "xlsx": 50_000,   # ~50 KB/page (not used — XLSX always goes to C++ probe)
    "pdf": 30_000,    # ~30 KB/page (not used — qpdf handles PDF)
}


def _estimate_pages_from_size(size_bytes: int, format: FormatName) -> int:
    """Estimate page count from file size when metadata is unavailable.

    Returns a conservative estimate (tends to over-count pages) so the
    chunk planner produces more, smaller chunks. This is safe because:
    - Over-chunking just means slightly more subprocess spawns
    - The adaptive chunk sizing will still produce reasonable chunk sizes
    - If a chunk covers pages that don't exist, the worker renders what's
      available (Aspose handles out-of-range page indices gracefully)

    Minimum return is 1 page.
    """
    bytes_per_page = _BYTES_PER_PAGE_ESTIMATE.get(format, 20_000)
    estimated = max(1, size_bytes // bytes_per_page)
    log.info(
        "probe_lite size-based estimate: format=%s size=%d bytes_per_page=%d estimated_pages=%d",
        format, size_bytes, bytes_per_page, estimated,
    )
    return estimated


async def _qpdf_page_count(path: Path) -> int | None:
    """qpdf --show-npages <file> → integer on stdout. Memory cost bounded.

    Returns None on non-zero exit or unparseable output (which then falls
    through to the C++ Aspose probe).
    """
    proc = await asyncio.create_subprocess_exec(
        "qpdf",
        "--show-npages",
        str(path),
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.DEVNULL,
    )
    stdout, _ = await proc.communicate()
    if proc.returncode != 0:
        return None
    m = _QPDF_NPAGES_RE.match(stdout)
    if m is None:
        return None
    try:
        return int(m.group(1))
    except ValueError:
        return None

File: internal/probe/probe_lite.go
package probe

import (
	"archive/zip"
	"context"
	"encoding/xml"
	"io"
	"os"
	"os/exec"
	"regexp"
	"strconv"
	"strings"

	"github.com/opus2/office-convert-orchestrator/internal/config"
	"github.com/opus2/office-convert-orchestrator/internal/oerrors"
	"github.com/opus2/office-convert-orchestrator/internal/types"
	"github.com/opus2/office-convert-orchestrator/internal/worker"
)

// staleMetadataSizeThreshold: a DOCX over this size claiming 1 page in app.xml
// is treated as stale and gets a size-based estimate instead.
const staleMetadataSizeThreshold = 200_000

// bytesPerPageEstimate — conservative (LOW) per-format byte-per-page figures
// for size-based fallback. Low => over-estimate pages => over-chunk (safe).
var bytesPerPageEstimate = map[types.FormatName]int64{
	types.FormatDOCX: 20_000,
	types.FormatPPTX: 100_000,
	types.FormatXLSX: 50_000,
	types.FormatPDF:  30_000,
}

var qpdfNPagesRE = regexp.MustCompile(`^\s*(\d+)\s*$`)

// ProbeLite is a best-effort metadata-only probe. Returns nil on any failure
// (caller falls through to the C++ Aspose worker). Mirrors probe_lite.probe_lite.
func ProbeLite(ctx context.Context, inputPath string, format types.FormatName) *types.ProbeResult {
	info, err := os.Stat(inputPath)
	if err != nil {
		return nil
	}
	size := info.Size()

	var n int
	switch format {
	case types.FormatDOCX:
		c := ooxmlCount(inputPath, "Pages")
		switch {
		case c <= 0:
			n = estimatePagesFromSize(size, format)
		case c == 1 && size > staleMetadataSizeThreshold:
			n = estimatePagesFromSize(size, format)
		default:
			n = c
		}
	case types.FormatPPTX:
		n = ooxmlCount(inputPath, "Slides")
		if n <= 0 {
			n = pptxSlideCountFromZip(inputPath)
		}
		if n <= 0 {
			n = estimatePagesFromSize(size, format)
		}
	case types.FormatXLSX:
		// XLSX always falls through to the C++ probe (rendered page count, not
		// worksheet count — see probe_lite.py).
		return nil
	case types.FormatPDF:
		n = qpdfPageCount(ctx, inputPath)
	default:
		return nil
	}

	if n <= 0 {
		return nil
	}
	return &types.ProbeResult{PageCount: n, Format: format, NaturalSeams: nil, SizeBytes: size}
}

// Probe is the two-tier probe: metadata-only first, full Aspose probe as
// fallback, with a format-mismatch retry. Mirrors probe.probe.
func Probe(ctx context.Context, s *config.Settings, inputPath string, format types.FormatName, requestID string) (types.ProbeResult, error) {
	if lite := ProbeLite(ctx, inputPath, format); lite != nil {
		return *lite, nil
	}

	stdout, _, err := worker.RunWorker(ctx, s, "probe", inputPath, format, "", nil, requestID, nil)
	if err == nil {
		return ParseProbeJSON(stdout)
	}

	// Format-mismatch retry: if the worker rejected the file with
	// input_unprocessable and the message hints a different format, retry once.
	if oe, ok := err.(*oerrors.Error); ok && oe.FailureClass == types.InputUnprocessable {
		if retry := mismatchRetryFormat(strings.ToLower(err.Error()), format); retry != "" {
			if stdout2, _, err2 := worker.RunWorker(ctx, s, "probe", inputPath, retry, "", nil, requestID, nil); err2 == nil {
				if pr, perr := ParseProbeJSON(stdout2); perr == nil {
					pr.Format = retry
					return pr, nil
				}
			}
		}
	}
	return types.ProbeResult{}, err
}

// mismatchRetryFormat reproduces probe.py's hint logic, including its Python
// operator-precedence behavior: the "excel"/"powerpoint" branches OR the
// keyword test with the (keyword AND format-differs) test.
func mismatchRetryFormat(errMsg string, format types.FormatName) types.FormatName {
	switch {
	case strings.Contains(errMsg, "word doc") && format != types.FormatDOCX:
		return types.FormatDOCX
	case strings.Contains(errMsg, "excel") || (strings.Contains(errMsg, "workbook") && format != types.FormatXLSX):
		return types.FormatXLSX
	case strings.Contains(errMsg, "powerpoint") || (strings.Contains(errMsg, "presentation") && format != types.FormatPPTX):
		return types.FormatPPTX
	}
	return ""
}

// ooxmlCount reads <Pages> or <Slides> from docProps/app.xml. Returns 0/-1 on
// any failure (treated as "no count").
func ooxmlCount(path, element string) int {
	zr, err := zip.OpenReader(path)
	if err != nil {
		return 0
	}
	defer zr.Close()
	for _, f := range zr.File {
		if f.Name != "docProps/app.xml" {
			continue
		}
		rc, err := f.Open()
		if err != nil {
			return 0
		}
		data, err := io.ReadAll(rc)
		rc.Close()
		if err != nil {
			return 0
		}
		var props struct {
			Pages  *int `xml:"Pages"`
			Slides *int `xml:"Slides"`
		}
		if err := xml.Unmarshal(data, &props); err != nil {
			return 0
		}
		if element == "Pages" && props.Pages != nil {
			return *props.Pages
		}
		if element == "Slides" && props.Slides != nil {
			return *props.Slides
		}
		return 0
	}
	return 0
}

// pptxSlideCountFromZip counts ppt/slides/slideN.xml entries.
func pptxSlideCountFromZip(path string) int {
	zr, err := zip.OpenReader(path)
	if err != nil {
		return 0
	}
	defer zr.Close()
	count := 0
	for _, f := range zr.File {
		if strings.HasPrefix(f.Name, "ppt/slides/slide") && strings.HasSuffix(f.Name, ".xml") {
			count++
		}
	}
	return count
}

func estimatePagesFromSize(size int64, format types.FormatName) int {
	bpp, ok := bytesPerPageEstimate[format]
	if !ok {
		bpp = 20_000
	}
	est := size / bpp
	if est < 1 {
		est = 1
	}
	return int(est)
}

// qpdfPageCount shells out to `qpdf --show-npages`. Returns 0 on any failure.
func qpdfPageCount(ctx context.Context, path string) int {
	out, err := exec.Command("qpdf", "--show-npages", path).Output()
	if err != nil {
		return 0
	}
	m := qpdfNPagesRE.FindSubmatch(out)
	if m == nil {
		return 0
	}
	n, err := strconv.Atoi(string(m[1]))
	if err != nil {
		return 0
	}
	return n
}
